triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -217,53 +220,70 @@ class Runner(
-                    applicationId = applicationId.unwrap,
-                    commandId = commandUUID.toString,
-                    party = party,
-                    commands = commands
-                  )
-                  logger.debug(
-                    s"submitting command ID $commandId, commands ${commands.map(_.command.value)}")
-                  submit(SubmitRequest(commands = Some(commandsArg)))
+  import Runner.{DamlTuple2, DamlFun}
+
+  @throws[RuntimeException]
+  private def handleCommands[Z](
+      commandId: String,
+      commands: Seq[Command],
+      submit: SubmitRequest => Z): Z = {
+    if (usedCommandIds.contains(commandId)) {
+      throw new RuntimeException(s"Duplicate command id: $commandId")
+    }
+    usedCommandIds += commandId
+    val commandUUID = UUID.randomUUID
+    commandIdMap += (commandUUID -> commandId)
+    val commandsArg = Commands(
+      ledgerId = client.ledgerId.unwrap,
+      applicationId = applicationId.unwrap,
+      commandId = commandUUID.toString,
+      party = party,
+      commands = commands
+    )
+    logger.debug(s"submitting command ID $commandId, commands ${commands.map(_.command.value)}")
+    submit(SubmitRequest(commands = Some(commandsArg)))
+  }
+
+  // Handles the value of update.
+  private def handleStepFreeResult(
+      clientTime: Timestamp,
+      v: SValue,
+      submit: SubmitRequest => Unit): SValue = {
+    def evaluate(se: SExpr) = {
+      val machine: Speedy.Machine = // TODO SC pull up
+        Speedy.Machine.fromPureSExpr(compiledPackages, se)
+      // Evaluate it.
+      machine.setExpressionToEvaluate(se)
+      Machine.stepToValue(machine)
+    }
+    def go(v: SValue): SValue =
+      unrollFree(v) match {
+        case Right(Right(vvv @ (variant, vv))) =>
+          vvv.match2 {
+            case "GetTime" /*(Time -> a)*/ => {
+              case DamlFun(timeA) => go(evaluate(makeAppD(timeA, STimestamp(clientTime))))
+            }
+            case "Submit" /*(Commands, () -> a)*/ => {
+              case DamlTuple2(commands, DamlFun(unitA)) =>
+                converter.toCommands(commands) match {
+                  case Left(err) => throw new ConverterException(err)
+                  case Right((commandId, commands)) =>
+                    handleCommands(commandId, commands, submit)
                 }
-              }
+                go(evaluate(makeAppD(unitA, SUnit)))
             }
-          case _ => {}
-        }
-        newState
-      }
-      case v => {
-        throw new RuntimeException(s"Expected Tuple2 but got $v")
+            case _ => {
+              case _ =>
+                val msg = s"unrecognized TriggerF step $variant"
+                logger.error(msg)
+                throw new ConverterException(msg)
+            }
+          }(fallback = throw new ConverterException(s"invalid contents for $variant: $vv"))
+        case Right(Left(newState)) => newState
+        case Left(e) => throw new RuntimeException(e)
```

Comment:
This was here before but a `ConverterException` might be more appropriate here.

Reply:
de429c7fa535509ed61b4ef0f98b1d6c31da786a